mullvad-relay-selector/tests/relay_selector.rs

```diff
@@ -0,0 +1,949 @@
+    }],
+    openvpn: OpenVpnEndpointData {
+        ports: vec![
+            OpenVpnEndpoint {
+                port: 1194,
+                protocol: TransportProtocol::Udp,
+            },
+            OpenVpnEndpoint {
+                port: 443,
+                protocol: TransportProtocol::Tcp,
+            },
+            OpenVpnEndpoint {
+                port: 80,
+                protocol: TransportProtocol::Tcp,
+            },
+        ],
+    },
+    bridge: BridgeEndpointData {
+        shadowsocks: vec![
+            ShadowsocksEndpointData {
+                port: 443,
+                cipher: "aes-256-gcm".to_string(),
+                password: "mullvad".to_string(),
+                protocol: TransportProtocol::Tcp,
+            },
+            ShadowsocksEndpointData {
+                port: 1234,
+                cipher: "aes-256-cfb".to_string(),
+                password: "mullvad".to_string(),
+                protocol: TransportProtocol::Udp,
+            },
+            ShadowsocksEndpointData {
+                port: 1236,
+                cipher: "aes-256-gcm".to_string(),
+                password: "mullvad".to_string(),
+                protocol: TransportProtocol::Udp,
+            },
+        ],
+    },
+    wireguard: WireguardEndpointData {
+        port_ranges: vec![
+            (53, 53),
+            (443, 443),
+            (4000, 33433),
+            (33565, 51820),
+            (52000, 60000),
+        ],
+        ipv4_gateway: "10.64.0.1".parse().unwrap(),
+        ipv6_gateway: "fc00:bbbb:bbbb:bb01::1".parse().unwrap(),
+        udp2tcp_ports: vec![],
+    },
+});
+
+// Some nifty constants.
+const UDP: TransportProtocol = TransportProtocol::Udp;
+const TCP: TransportProtocol = TransportProtocol::Tcp;
+
+// Helper functions
+fn unwrap_relay(get_result: GetRelay) -> Relay {
+    match get_result {
+        GetRelay::Wireguard { inner, .. } => match inner {
+            crate::WireguardConfig::Singlehop { exit } => exit,
+            crate::WireguardConfig::Multihop { exit, .. } => exit,
+        },
+        GetRelay::OpenVpn { exit, .. } => exit,
+        GetRelay::Custom(custom) => {
+            panic!("Can not extract regular relay from custom relay: {custom}")
+        }
+    }
+}
+
+fn unwrap_endpoint(get_result: GetRelay) -> MullvadEndpoint {
+    match get_result {
+        GetRelay::Wireguard { endpoint, .. } => endpoint,
+        GetRelay::OpenVpn { endpoint, .. } => endpoint,
+        GetRelay::Custom(custom) => {
+            panic!("Can not extract Mullvad endpoint from custom relay: {custom}")
+        }
+    }
+}
+
+fn tunnel_type(relay: &Relay) -> TunnelType {
+    match relay.endpoint_data {
+        RelayEndpointData::Openvpn | RelayEndpointData::Bridge => TunnelType::OpenVpn,
+        RelayEndpointData::Wireguard(_) => TunnelType::Wireguard,
+    }
+}
+
+/// Test whether the relay selector seems to respect the order as defined by [`RETRY_ORDER`].
+#[test]
+fn test_retry_order() {
```

Comment:
It could, but I struggle to come up with properties that are valid across all configurations of `RETRY_ORDER`. I'm all for adding properties to test for if it can be done generically, or add more specialized test cases that may break if `RETRY_ORDER` is updated :blush:

Reply:
Nice! Updated the test, but it still omits some properties of the chosen relays (i.e. multihop/bridges). What do you think?

Reply:
I see that you added a test for bridges. LGTM! 💯